test/integration/apiserver/cel/mutatingadmissionpolicy_test.go

```diff
@@ -0,0 +1,925 @@
+	coreinstall "k8s.io/kubernetes/pkg/apis/core/install"
+	"k8s.io/kubernetes/test/integration/framework"
+	"k8s.io/utils/ptr"
+)
+
+func policy(name string, matchResources v1alpha1.MatchResources, paramKind *v1alpha1.ParamKind, mutations ...v1alpha1.Mutation) *v1alpha1.MutatingAdmissionPolicy {
+	return &v1alpha1.MutatingAdmissionPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: v1alpha1.MutatingAdmissionPolicySpec{
+			MatchConstraints: &matchResources,
+			ParamKind:        paramKind,
+			Mutations:        mutations,
+		},
+	}
+}
+
+func binding(policyName string, paramRef *v1alpha1.ParamRef, matchResources *v1alpha1.MatchResources) *v1alpha1.MutatingAdmissionPolicyBinding {
+	return &v1alpha1.MutatingAdmissionPolicyBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: policyName + "-binding-" + string(uuid.NewUUID()),
+		},
+		Spec: v1alpha1.MutatingAdmissionPolicyBindingSpec{
+			PolicyName:     policyName,
+			ParamRef:       paramRef,
+			MatchResources: matchResources,
+		},
+	}
+}
+
+func TestMAP(t *testing.T) {
+	matchAllResources := v1alpha1.MatchResources{
+		ResourceRules: []v1alpha1.NamedRuleWithOperations{
+			{
+				RuleWithOperations: v1alpha1.RuleWithOperations{
+					Operations: []admissionregistrationv1.OperationType{"*"},
+					Rule: admissionregistrationv1.Rule{
+						APIGroups:   []string{"*"},
+						APIVersions: []string{"*"},
+						Resources:   []string{"*"},
+					},
+				},
+			},
+		},
+	}
+
+	cases := []struct {
+		name     string
+		policies []*v1alpha1.MutatingAdmissionPolicy
+		bindings []*v1alpha1.MutatingAdmissionPolicyBinding
+		params   []runtime.Object
+
+		requestOperation admissionregistrationv1.OperationType
+		requestResource  schema.GroupVersionResource
+		requestObject    runtime.Object
+
+		expected runtime.Object
+	}{
+		{
+			name: "basic",
+			policies: []*v1alpha1.MutatingAdmissionPolicy{
+				policy("basic-policy", matchAllResources, nil, v1alpha1.Mutation{
+					PatchType:          ptr.To(v1alpha1.ApplyConfigurationPatchType),
+					ReinvocationPolicy: ptr.To(v1alpha1.NeverReinvocationPolicy),
+					Expression: `
+						Object{
+							metadata: Object.metadata{
+								annotations: {
+									"my-foo-annotation": "myAnnotationValue"
+								}
+							}
+						}`,
+				}),
+			},
+			bindings: []*v1alpha1.MutatingAdmissionPolicyBinding{
+				binding("basic-policy", nil, nil),
+			},
+			requestOperation: admissionregistrationv1.Create,
+			requestResource:  corev1.SchemeGroupVersion.WithResource("configmaps"),
+			requestObject: &corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-configmap",
+					Namespace: "default",
+				},
+			},
+			expected: &corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-configmap",
+					Namespace: "default",
+					Annotations: map[string]string{
+						"my-foo-annotation": "myAnnotationValue",
+					},
+				},
+			},
+		},
+		{
+			name: "multiple policies",
```

Comment:
Can we add a test with reinvocation policy `IfNeeded` where one policy adds a list of objects, and another policy sets a field on each object in the list.  With reinvocation that should always succeed.